docs/decisions/0004-error-handling.md

```diff
@@ -0,0 +1,59 @@
+
+- **Other**:
+  - Remove any code that wraps unhandled exceptions into AIException or any other SK exception solely for the purpose of wrapping. In most cases, this code does not provide useful information to help SK client code handle the exception, apart from a generic and uninformative "Something went wrong" message. Regardless of whether the unhandled exception is wrapped or not, SK client code cannot handle or recover from it without prior knowledge of the exception and proper exception handling logic.
+
+## Questions and consideration
+
+- Should the exception hierarchy follow the SK components hierarchy? For example, should there be a PlanningException for planners and a MemoryStorageException for memory storage connectors? What are the benefits for SK client code? Can the SK client code operate on the exception?  
+  In order to handle an exception successfully, the client code needs to know as many details as possible about the failed operation, such as exception type, status code, and some other operation related properties. This suggests that specific exception types with strongly typed properties should be preferred over generic/common exception types. If that's the case, what is the purpose of having base generic exceptions?  
+  On the other hand, having base exception types per component or set of related operations can be convenient from a code reusability perspective because derived types can inherit common component exception properties relevant to that component.
+- Should all custom SK exceptions be derived from the SKException type?  Currently, SKException derives from the Exception type and does not add any additional information. What value can SK client code get from catching these exceptions compared to catching them as Exception types? Should SKException be used in all cases when SK needs to indicate that something went wrong without providing details through strongly typed properties, and neither custom nor standard exception types exist that could be used for that purpose?
+- Should we keep the IsCriticalException functionality that SK currently has to avoid handling several system/critical exceptions such as OutOfMemoryException and StackOverflowException? Some of them cannot be intercepted.
+- An idea has been proposed to add SK wide error codes to exceptions in order to identify and locate the source of the exception. The question is, is it necessary? Aren't the exception type, message, values of its properties, and stack trace enough to precisely identify the exception and locate its source?
```

Comment:
I think it depends how granular our exception hierarchy will be. Error codes are usually short, and they are useful when you try to search specific error in your logs (although you can still search by exception type), or if you try to find specific problem in documentation (again, still searchable by exception type).

The other point is - error codes will allow to align exception strategy across different programming languages (C#, Python, Java and more in the future). While exception hierarchy can be designed with inheritance in those languages, the naming convention for exceptions might be different (for example `MemoryStoreException` in C# and `MemoryStoreError` in Python). At the same time, error code like `SK001` can point to the same problem in all languages.

Do we actually need to align exceptions across different languages? I believe that is separate topic to discuss :)

Reply:
Cross-language alignment may seem good, but it can come at a cost of unnecessary coordination and maintenance, slowing down the process and teams involved.